extensions/amp-user-notification/0.1/amp-user-notification.js

```diff
@@ -0,0 +1,378 @@
+      this.userNotificationManager_ = getUserNotificationManager_(this.win_);
+
+      this.elementId_ = assert(this.element.id,
+          'amp-user-notification should have an id.');
+
+      assert(this.element.hasAttribute('data-show-if-href'),
+          `"amp-user-notification" (${this.elementId_}) ` +
+          'should have "data-show-if-href" attribute.');
+      /** @private @const {string} */
+      this.showIfHref_ = assertHttpsUrl(
+          this.element.getAttribute('data-show-if-href'), this.element);
+
+      assert(this.element.hasAttribute('data-dismiss-href'),
+          `"amp-user-notification" (${this.elementId_}) ` +
+          'should have "data-dismiss-href" attribute.');
+
+      /** @private @const {string} */
+      this.dismissHref_ = assertHttpsUrl(
+          this.element.getAttribute('data-dismiss-href'), this.element);
+
+      this.userNotificationManager_
+          .registerUserNotification(this.elementId_, this);
+
+      this.registerAction('dismiss', this.dismiss.bind(this));
+    }
+  }
+
+  /**
+   * Executes a `POST` request to the url given on the `data-show-if-href`
+   * attribute.
+   * @param {string} ampUserId
+   * @return {!Promise<!PostResponseMetadataDef>}
+   * @private
+   */
+  postShowEndpoint_(ampUserId) {
+    // TODO(erwinm, https://github.com/ampproject/amphtml/issues/1228)
+    // Determine wether this should be a POST with json body
+    // or a GET with data on the query params.
+    this.ampUserId_ = ampUserId;
+    const getReq = {
+      method: 'POST',
+      credentials: 'include',
+      /** @type {PostRequestMetadataDef} */
+      body: {
+        'elementId': this.elementId_,
+        'ampUserId': ampUserId
+      }
+    };
+
+    return xhrFor(this.win_).fetchJson(this.showIfHref_, getReq);
+  }
+
+  /**
+   * Creates an POST to the specified `data-dismiss-href` url.
+   * @private
+   * @return {!Promise}
+   */
+  postDismissEnpoint_() {
+    return xhrFor(this.win_).fetchJson(this.dismissHref_, {
+      method: 'POST',
+      credentials: 'include',
+      body: {
+        'elementId': this.elementId_,
+        'ampUserId': this.ampUserId_
+      }
+    });
+  }
+
+  /**
+   * Success handler for `postShowEndpoint_`.
+   * @param {!PostResponseMetadataDef}
+   * @return {!Promise<boolean>}
+   * @private
+   */
+  onPostShowEndpointSuccess_(data) {
+    assert(typeof data['showNotification'] == 'boolean', '`showNotification` ' +
+        'should be a boolean. Got "%s" which is of type %s.',
+        data['showNotification'], typeof data['showNotification']);
+
+    if (!data['showNotification']) {
+      // If no notification needs to be shown, resolve the `dialogPromise_`
+      // right away with false.
+      this.dialogResolve_();
+    }
+    return Promise.resolve(data['showNotification']);
+  }
+
+  /**
+   * Get async cid service.
+   * @return {!Promise}
+   * @private
+   */
+  getAsyncCid_() {
+    this.dialogPromise_ = new Promise(resolve => {
```

Comment:
It looks like this can be moved up in the constructor. There's no cost in creating it right away. But that would improve readability because `this.dialogResolve_` is used throughout above without assert protection. Putting it in the constructor will make this clear.